content/systems/filecoin_files/piece/_index.md

```diff
@@ -31,7 +31,7 @@ It is important to highlight that data submitted to the Filecoin network go thro
 
 1. When a piece of data, or file is submitted to Filecoin (in some raw system format) it is transformed into a _UnixFS DAG style data representation_ (in case it is not in this format already, e.g., from IPFS-based applications). The hash that represents the root of the IPLD DAG of the UnixFS file is the _Payload CID_, which is used in the Retrieval Market. The Payload CID is identical to an IPFS CID.
 2. In order to make a _Filecoin Piece_ the UnixFS IPLD DAG is serialised into a ["Content-Addressable aRchive" (.car)](https://github.com/ipld/specs/blob/master/block-layer/content-addressable-archives.md#summary) file, which is in raw bytes format.
-3. The resulting .car file is _padded_ with extra bits.
+3. The resulting .car file is _padded_ with extra bits in order to get it to "power of 2" size. This is done in order for the file to make a binary Merkle tree. This means that two zero (0) bits need to be added to every 254 bits (to make the 256 bits). In case more padding is needed in order to reach the 254 bit threshold, then these bits are also filled with zeros.
 4. The next step is to calculate the Merkle root out of the hashes of the Piece. The resulting root of the Merkle tree is the **Piece CID**. This is also referred to as _CommP_ or _Piece Commitment_.
```

Comment:
This description isn't clear to me. The bit padding (the two zero bits after every 254 bits) don't really have anything to do with the 'power of 2' padding. They're better thought of as two distinct, sequential steps. First the bit padding is added. Then the power-of-2 padding is added.

The description as written describes the motivation for bit padding as 'in order to get it to "power of 2" size' — but that's not the case. So the text above never really gets around to describing the actual power of two padding.

I don't understand what the last sentence of 3. is trying to communicate. It seems either to be a repetition of the previous sentence's content — or else meant to be a return to the promise of power of two padding. But if the latter, it doesn't actually describe that at all.

Reply:
OK, so @porcuquine is almost certainly a better way of explaining it to communicate the concepts better! In code, the steps are:

1. find the next `^2 x 254/256` size above the input size
2. pad out the input to that size
3. pass through fr32 padding to add those extra 2 bits per 254 bytes

That first step is the most difficult to describe because it involves multiple parts and is just weird.

_BUT_ because we're dealing with zeros here, it can be flipped and the fr32 can be done beforehand, this code flow is probably just an artifact of the various streaming pieces or some legacy. The fr32 padding needs to be done across the entire piece, but since the byte padding is zero bytes anyway there's no good reason to require it be done as a last part of the step.

1. fr32 pad, adding 2 bits per 254 bits
2. figure out the next ^2 greater than the size after step 1
3. add zero bytes to pad input out to that length

Reply:
Yes, what @rvagg wrote above is correct. Sorry for the late response.

What I was describing corresponds to his second set of three steps. I think it's much simpler to understand.
That said, @rvagg is correct that the more complicated case is what is actually implemented, and  `rust-fil-proofs` does check the size before adding the bit-padding (by which I mean 2 zero bits after every 254 of raw input).

It's my belief that the spec only has to and only *should* describe the result — not the steps followed to produce it. So I would favor the simpler explanation, even though it doesn't correspond to the work flow imposed by the `rust-fil-proofs` API.

If you do want to use the other set of three steps (his original ones), then the description will need to find a way to articulate the hard-to-describe step 1.

Also missing entirely here is the minimum piece size, which is defined to be 127 bytes before bit padding (or 128 bytes after bit padding). For gory detail, see: https://github.com/filecoin-project/rust-fil-proofs/issues/1231

Reply:
I think you're a better judge of the audience here so should decide what level of detail to include. "in order for the file to make a binary Merkle tree" is only correct for one stage of padding, the fr32 is something else all-together. I'm not sure for this audience how much it matters that "your input padded internally and at the length such that it doesn't quite resemble your input before it is merklized". For now maybe it's fine as is, implementers and anyone else wanting to generate CommP separate from a full client (like we've had to do for these offline deals for Filecoin Discover data) will have to come to a more sophisticated understanding of the process, and this might not be the place for them to get that detail! Your call @yiannisbot, I think you have most of the details now.

Reply:
>and therefore I'm keen to go back and just have that: "The resulting .car file is padded with extra bits in order to get it to "power of 2" size. This is done in order for the file to make a binary Merkle tree.", which is the essence of it.

I understand the impulse, but this description does not provide enough information to know *what* must be done regardless of the order of steps — so it's not equivalent.

> If we want the process too, then this is the simplest explanation IMO:

>> fr32 pad, adding 2 bits per 254 bits
>> figure out the next ^2 greater than the size after step 1
>> add zero bytes to pad input out to that length

I agree. I think you could use that, or a similar description, perhaps with some framing to make clear that the steps are descriptive of the result, not prescriptive of the algorithm/API which must be used/provided.